// gg-editor/cjs/components/ItemPanel/Item.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

require("core-js/modules/es.object.define-property");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = void 0;

var _objectSpread2 = _interopRequireDefault(require("@babel/runtime/helpers/objectSpread"));

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _react = _interopRequireDefault(require("react"));

var _withGGEditorContext = _interopRequireDefault(require("../../common/context/GGEditorContext/withGGEditorContext"));

var Item =
/*#__PURE__*/
function (_React$Component) {
  (0, _inherits2["default"])(Item, _React$Component);

  function Item(props) {
    var _this;

    (0, _classCallCheck2["default"])(this, Item);
    _this = (0, _possibleConstructorReturn2["default"])(this, (0, _getPrototypeOf2["default"])(Item).call(this, props));

    _this.handleMouseDown = function () {
      var _this$props = _this.props,
          type = _this$props.type,
          size = _this$props.size,
          shape = _this$props.shape,
          model = _this$props.model;

      if (_this.page) {
        _this.page.beginAdd(type, (0, _objectSpread2["default"])({
          type: type,
          size: size,
          shape: shape
        }, model));
      }
    };

    _this.bindEvent();

    return _this;
  }

  (0, _createClass2["default"])(Item, [{
    key: "bindEvent",
    value: function bindEvent() {
      var _this2 = this;

      var onAfterAddPage = this.props.onAfterAddPage;
      onAfterAddPage(function (_ref) {
        var page = _ref.page;
        _this2.page = page;
      });
    }
  }, {
    key: "render",
    value: function render() {
      var _this$props2 = this.props,
          src = _this$props2.src,
          shape = _this$props2.shape,
          children = _this$props2.children;
      return _react["default"].createElement("div", {
        style: {
          cursor: 'pointer'
        },
        onMouseDown: this.handleMouseDown
      }, src ? _react["default"].createElement("img", {
        src: src,
        alt: shape,
        draggable: false
      }) : children);
    }
  }]);
  return Item;
}(_react["default"].Component);

var _default = (0, _withGGEditorContext["default"])(Item);

exports["default"] = _default;
