// @antv/l7-utils/lib/fetchData.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.getImage = exports.getArrayBuffer = exports.getJSON = void 0;

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _wrapNativeSuper2 = _interopRequireDefault(require("@babel/runtime/helpers/wrapNativeSuper"));

function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = (0, _getPrototypeOf2.default)(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = (0, _getPrototypeOf2.default)(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return (0, _possibleConstructorReturn2.default)(this, result); }; }

function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Boolean.prototype.valueOf.call(Reflect.construct(Boolean, [], function () {})); return true; } catch (e) { return false; } }

var AJAXError = function (_Error) {
  (0, _inherits2.default)(AJAXError, _Error);

  var _super = _createSuper(AJAXError);

  function AJAXError(message, status, url) {
    var _this;

    (0, _classCallCheck2.default)(this, AJAXError);
    _this = _super.call(this, message);
    _this.status = void 0;
    _this.url = void 0;
    _this.status = status;
    _this.url = url;
    _this.name = _this.constructor.name;
    _this.message = message;
    return _this;
  }

  (0, _createClass2.default)(AJAXError, [{
    key: "toString",
    value: function toString() {
      return "".concat(this.name, ": ").concat(this.message, " (").concat(this.status, "): ").concat(this.url);
    }
  }]);
  return AJAXError;
}((0, _wrapNativeSuper2.default)(Error));

function makeRequest(requestParameters) {
  var xhr = new XMLHttpRequest();
  xhr.open('GET', requestParameters.url, true);

  for (var k in requestParameters.headers) {
    if (requestParameters.headers.hasOwnProperty(k)) {
      xhr.setRequestHeader(k, requestParameters.headers[k]);
    }
  }

  xhr.withCredentials = requestParameters.credentials === 'include';
  return xhr;
}

var getJSON = function getJSON(requestParameters, callback) {
  var xhr = makeRequest(requestParameters);
  xhr.setRequestHeader('Accept', 'application/json');

  xhr.onerror = function () {
    callback(new Error(xhr.statusText));
  };

  xhr.onload = function () {
    if (xhr.status >= 200 && xhr.status < 300 && xhr.response) {
      var data;

      try {
        data = JSON.parse(xhr.response);
      } catch (err) {
        return callback(err);
      }

      callback(null, data);
    } else {
      if (xhr.status === 401) {
        callback(new AJAXError("".concat(xhr.statusText), xhr.status, requestParameters.url));
      } else {
        callback(new AJAXError(xhr.statusText, xhr.status, requestParameters.url));
      }
    }
  };

  xhr.send();
  return xhr;
};

exports.getJSON = getJSON;

var getArrayBuffer = function getArrayBuffer(requestParameters, callback) {
  var xhr = makeRequest(requestParameters);
  xhr.responseType = 'arraybuffer';

  xhr.onerror = function () {
    callback(new Error(xhr.statusText));
  };

  xhr.onload = function () {
    var response = xhr.response;

    if (response.byteLength === 0 && xhr.status === 200) {
      return callback(new Error('http status 200 returned without content.'));
    }

    if (xhr.status >= 200 && xhr.status < 300 && xhr.response) {
      callback(null, {
        data: response,
        cacheControl: xhr.getResponseHeader('Cache-Control'),
        expires: xhr.getResponseHeader('Expires')
      });
    } else {
      callback(new AJAXError(xhr.statusText, xhr.status, requestParameters.url));
    }
  };

  xhr.send();
  return xhr;
};

exports.getArrayBuffer = getArrayBuffer;

function sameOrigin(url) {
  var a = window.document.createElement('a');
  a.href = url;
  return a.protocol === window.document.location.protocol && a.host === window.document.location.host;
}

var transparentPngUrl = 'data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAAEAAAABCAYAAAAfFcSJAAAAC0lEQVQYV2NgAAIAAAUAAarVyFEAAAAASUVORK5CYII=';

var getImage = function getImage(requestParameters, callback) {
  return getArrayBuffer(requestParameters, function (err, imgData) {
    if (err) {
      callback(err);
    } else if (imgData) {
      var img = new window.Image();
      img.crossOrigin = 'anonymous';
      var URL = window.URL || window.webkitURL;

      img.onload = function () {
        callback(null, img);
        URL.revokeObjectURL(img.src);
      };

      var blob = new window.Blob([new Uint8Array(imgData.data)], {
        type: 'image/png'
      });
      img.src = imgData.data.byteLength ? URL.createObjectURL(blob) : transparentPngUrl;
    }
  });
};

exports.getImage = getImage;
