// @antv/l7-core/lib/services/scene/SceneService.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _regenerator = _interopRequireDefault(require("@babel/runtime/regenerator"));

var _asyncToGenerator2 = _interopRequireDefault(require("@babel/runtime/helpers/asyncToGenerator"));

var _initializerDefineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/initializerDefineProperty"));

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _assertThisInitialized2 = _interopRequireDefault(require("@babel/runtime/helpers/assertThisInitialized"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _applyDecoratedDescriptor2 = _interopRequireDefault(require("@babel/runtime/helpers/applyDecoratedDescriptor"));

var _initializerWarningHelper2 = _interopRequireDefault(require("@babel/runtime/helpers/initializerWarningHelper"));

var _asyncHook = require("@antv/async-hook");

var _l7Utils = require("@antv/l7-utils");

var _elementResizeEvent = _interopRequireWildcard(require("element-resize-event"));

var _eventemitter = require("eventemitter3");

var _inversify = require("inversify");

var _types = require("../../types");

var _dom = require("../../utils/dom");

var _IInteractionService = require("../interaction/IInteractionService");

var _dec, _dec2, _dec3, _dec4, _dec5, _dec6, _dec7, _dec8, _dec9, _dec10, _dec11, _dec12, _dec13, _dec14, _dec15, _dec16, _dec17, _class, _class2, _descriptor, _descriptor2, _descriptor3, _descriptor4, _descriptor5, _descriptor6, _descriptor7, _descriptor8, _descriptor9, _descriptor10, _descriptor11, _descriptor12, _descriptor13, _descriptor14, _descriptor15, _descriptor16, _temp;

function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = (0, _getPrototypeOf2.default)(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = (0, _getPrototypeOf2.default)(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return (0, _possibleConstructorReturn2.default)(this, result); }; }

function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Boolean.prototype.valueOf.call(Reflect.construct(Boolean, [], function () {})); return true; } catch (e) { return false; } }

var Scene = (_dec = (0, _inversify.injectable)(), _dec2 = (0, _inversify.inject)(_types.TYPES.SceneID), _dec3 = (0, _inversify.inject)(_types.TYPES.IIconService), _dec4 = (0, _inversify.inject)(_types.TYPES.IFontService), _dec5 = (0, _inversify.inject)(_types.TYPES.IControlService), _dec6 = (0, _inversify.inject)(_types.TYPES.ILogService), _dec7 = (0, _inversify.inject)(_types.TYPES.IGlobalConfigService), _dec8 = (0, _inversify.inject)(_types.TYPES.IMapService), _dec9 = (0, _inversify.inject)(_types.TYPES.ICoordinateSystemService), _dec10 = (0, _inversify.inject)(_types.TYPES.IRendererService), _dec11 = (0, _inversify.inject)(_types.TYPES.ILayerService), _dec12 = (0, _inversify.inject)(_types.TYPES.ICameraService), _dec13 = (0, _inversify.inject)(_types.TYPES.IInteractionService), _dec14 = (0, _inversify.inject)(_types.TYPES.IPickingService), _dec15 = (0, _inversify.inject)(_types.TYPES.IShaderModuleService), _dec16 = (0, _inversify.inject)(_types.TYPES.IMarkerService), _dec17 = (0, _inversify.inject)(_types.TYPES.IPopupService), _dec(_class = (_class2 = (_temp = function (_EventEmitter) {
  (0, _inherits2.default)(Scene, _EventEmitter);

  var _super = _createSuper(Scene);

  function Scene() {
    var _this;

    (0, _classCallCheck2.default)(this, Scene);
    _this = _super.call(this);
    _this.destroyed = false;
    _this.loaded = false;
    (0, _initializerDefineProperty2.default)(_this, "id", _descriptor, (0, _assertThisInitialized2.default)(_this));
    (0, _initializerDefineProperty2.default)(_this, "iconService", _descriptor2, (0, _assertThisInitialized2.default)(_this));
    (0, _initializerDefineProperty2.default)(_this, "fontService", _descriptor3, (0, _assertThisInitialized2.default)(_this));
    (0, _initializerDefineProperty2.default)(_this, "controlService", _descriptor4, (0, _assertThisInitialized2.default)(_this));
    (0, _initializerDefineProperty2.default)(_this, "logger", _descriptor5, (0, _assertThisInitialized2.default)(_this));
    (0, _initializerDefineProperty2.default)(_this, "configService", _descriptor6, (0, _assertThisInitialized2.default)(_this));
    (0, _initializerDefineProperty2.default)(_this, "map", _descriptor7, (0, _assertThisInitialized2.default)(_this));
    (0, _initializerDefineProperty2.default)(_this, "coordinateSystemService", _descriptor8, (0, _assertThisInitialized2.default)(_this));
    (0, _initializerDefineProperty2.default)(_this, "rendererService", _descriptor9, (0, _assertThisInitialized2.default)(_this));
    (0, _initializerDefineProperty2.default)(_this, "layerService", _descriptor10, (0, _assertThisInitialized2.default)(_this));
    (0, _initializerDefineProperty2.default)(_this, "cameraService", _descriptor11, (0, _assertThisInitialized2.default)(_this));
    (0, _initializerDefineProperty2.default)(_this, "interactionService", _descriptor12, (0, _assertThisInitialized2.default)(_this));
    (0, _initializerDefineProperty2.default)(_this, "pickingService", _descriptor13, (0, _assertThisInitialized2.default)(_this));
    (0, _initializerDefineProperty2.default)(_this, "shaderModuleService", _descriptor14, (0, _assertThisInitialized2.default)(_this));
    (0, _initializerDefineProperty2.default)(_this, "markerService", _descriptor15, (0, _assertThisInitialized2.default)(_this));
    (0, _initializerDefineProperty2.default)(_this, "popupService", _descriptor16, (0, _assertThisInitialized2.default)(_this));
    _this.inited = false;
    _this.initPromise = void 0;
    _this.rendering = false;
    _this.$container = void 0;
    _this.canvas = void 0;
    _this.markerContainer = void 0;
    _this.hooks = void 0;

    _this.handleWindowResized = function () {
      _this.emit('resize');

      if (_this.$container) {
        _this.initContainer();

        _l7Utils.DOM.triggerResize();

        _this.coordinateSystemService.needRefresh = true;

        _this.render();
      }
    };

    _this.handleMapCameraChanged = function (viewport) {
      _this.cameraService.update(viewport);

      _this.render();
    };

    _this.hooks = {
      init: new _asyncHook.AsyncParallelHook()
    };
    return _this;
  }

  (0, _createClass2.default)(Scene, [{
    key: "init",
    value: function init(sceneConfig) {
      var _this2 = this;

      this.configService.setSceneConfig(this.id, sceneConfig);
      this.shaderModuleService.registerBuiltinModules();
      this.iconService.init();
      this.fontService.init();
      this.hooks.init.tapPromise('initMap', (0, _asyncToGenerator2.default)(_regenerator.default.mark(function _callee() {
        return _regenerator.default.wrap(function _callee$(_context) {
          while (1) {
            switch (_context.prev = _context.next) {
              case 0:
                _context.next = 2;
                return new Promise(function (resolve) {
                  _this2.map.onCameraChanged(function (viewport) {
                    _this2.cameraService.init();

                    _this2.cameraService.update(viewport);

                    resolve();
                  });

                  _this2.map.init();
                });

              case 2:
                _this2.map.onCameraChanged(_this2.handleMapCameraChanged);

                _this2.map.addMarkerContainer();

                _this2.markerService.addMarkers();

                _this2.markerService.addMarkerLayers();

                _this2.popupService.initPopup();

                _this2.interactionService.init();

                _this2.interactionService.on(_IInteractionService.InteractionEvent.Drag, _this2.addSceneEvent.bind(_this2));

                _this2.logger.debug("map ".concat(_this2.id, " loaded"));

              case 10:
              case "end":
                return _context.stop();
            }
          }
        }, _callee);
      })));
      this.hooks.init.tapPromise('initRenderer', (0, _asyncToGenerator2.default)(_regenerator.default.mark(function _callee2() {
        var $container;
        return _regenerator.default.wrap(function _callee2$(_context2) {
          while (1) {
            switch (_context2.prev = _context2.next) {
              case 0:
                $container = (0, _dom.createRendererContainer)(_this2.configService.getSceneConfig(_this2.id).id || '');
                _this2.$container = $container;

                if (!$container) {
                  _context2.next = 11;
                  break;
                }

                _this2.canvas = _l7Utils.DOM.create('canvas', '', $container);

                _this2.setCanvas();

                _context2.next = 7;
                return _this2.rendererService.init(_this2.canvas, _this2.configService.getSceneConfig(_this2.id));

              case 7:
                (0, _elementResizeEvent.default)(_this2.$container, _this2.handleWindowResized);
                window.matchMedia('screen and (-webkit-min-device-pixel-ratio: 1.5)').addListener(_this2.handleWindowResized);
                _context2.next = 12;
                break;

              case 11:
                _this2.logger.error('容器 id 不存在');

              case 12:
                _this2.pickingService.init(_this2.id);

                _this2.logger.debug("scene ".concat(_this2.id, " renderer loaded"));

              case 14:
              case "end":
                return _context2.stop();
            }
          }
        }, _callee2);
      })));
      this.initPromise = this.hooks.init.promise();
      this.render();
    }
  }, {
    key: "addLayer",
    value: function addLayer(layer) {
      this.logger.debug("add layer ".concat(layer.name, " to scene ").concat(this.id));
      this.layerService.add(layer);
      this.render();
    }
  }, {
    key: "render",
    value: function () {
      var _render = (0, _asyncToGenerator2.default)(_regenerator.default.mark(function _callee3() {
        return _regenerator.default.wrap(function _callee3$(_context3) {
          while (1) {
            switch (_context3.prev = _context3.next) {
              case 0:
                if (!(this.rendering || this.destroyed)) {
                  _context3.next = 2;
                  break;
                }

                return _context3.abrupt("return");

              case 2:
                this.rendering = true;

                if (this.inited) {
                  _context3.next = 13;
                  break;
                }

                _context3.next = 6;
                return this.initPromise;

              case 6:
                if (this.destroyed) {
                  this.destroy();
                }

                this.logger.info(' render inited');
                this.layerService.initLayers();
                this.controlService.addControls();
                this.loaded = true;
                this.emit('loaded');
                this.inited = true;

              case 13:
                this.layerService.renderLayers();
                this.logger.debug("scene ".concat(this.id, " render"));
                this.rendering = false;

              case 16:
              case "end":
                return _context3.stop();
            }
          }
        }, _callee3, this);
      }));

      function render() {
        return _render.apply(this, arguments);
      }

      return render;
    }()
  }, {
    key: "getSceneContainer",
    value: function getSceneContainer() {
      return this.$container;
    }
  }, {
    key: "exportPng",
    value: function exportPng(type) {
      var _this$$container;

      var renderCanvas = (_this$$container = this.$container) === null || _this$$container === void 0 ? void 0 : _this$$container.getElementsByTagName('canvas')[0];
      this.render();
      var layersPng = type === 'jpg' ? renderCanvas === null || renderCanvas === void 0 ? void 0 : renderCanvas.toDataURL('image/jpeg') : renderCanvas === null || renderCanvas === void 0 ? void 0 : renderCanvas.toDataURL('image/png');
      return layersPng;
    }
  }, {
    key: "getSceneConfig",
    value: function getSceneConfig() {
      return this.configService.getSceneConfig(this.id);
    }
  }, {
    key: "addMarkerContainer",
    value: function addMarkerContainer() {
      var mapContainer = this.$container.parentElement;

      if (mapContainer !== null) {
        this.markerContainer = _l7Utils.DOM.create('div', 'l7-marker-container', mapContainer);
      }
    }
  }, {
    key: "getMarkerContainer",
    value: function getMarkerContainer() {
      return this.markerContainer;
    }
  }, {
    key: "destroy",
    value: function destroy() {
      if (!this.inited) {
        this.destroyed = true;
        return;
      }

      this.emit('destroy');
      this.layerService.destroy();
      this.rendererService.destroy();
      this.map.destroy();
      this.interactionService.destroy();
      this.controlService.destroy();
      this.markerService.destroy();
      this.removeAllListeners();
      this.inited = false;
      (0, _elementResizeEvent.unbind)(this.$container, this.handleWindowResized);
      window.matchMedia('screen and (min-resolution: 2dppx)').removeListener(this.handleWindowResized);
    }
  }, {
    key: "initContainer",
    value: function initContainer() {
      var _this$$container2, _this$$container3;

      var pixelRatio = _l7Utils.DOM.DPR;
      var w = ((_this$$container2 = this.$container) === null || _this$$container2 === void 0 ? void 0 : _this$$container2.clientWidth) || 400;
      var h = ((_this$$container3 = this.$container) === null || _this$$container3 === void 0 ? void 0 : _this$$container3.clientHeight) || 300;
      var canvas = this.canvas;

      if (canvas) {
        canvas.width = w * pixelRatio;
        canvas.height = h * pixelRatio;
        canvas.style.width = "".concat(w, "px");
        canvas.style.height = "".concat(h, "px");
      }

      this.rendererService.viewport({
        x: 0,
        y: 0,
        width: pixelRatio * w,
        height: pixelRatio * h
      });
    }
  }, {
    key: "setCanvas",
    value: function setCanvas() {
      var _this$$container4, _this$$container5;

      var pixelRatio = _l7Utils.DOM.DPR;
      var w = ((_this$$container4 = this.$container) === null || _this$$container4 === void 0 ? void 0 : _this$$container4.clientWidth) || 400;
      var h = ((_this$$container5 = this.$container) === null || _this$$container5 === void 0 ? void 0 : _this$$container5.clientHeight) || 300;
      var canvas = this.canvas;
      canvas.width = w * pixelRatio;
      canvas.height = h * pixelRatio;
      canvas.style.width = "".concat(w, "px");
      canvas.style.height = "".concat(h, "px");
    }
  }, {
    key: "addSceneEvent",
    value: function addSceneEvent(target) {
      this.emit(target.type, target);
    }
  }]);
  return Scene;
}(_eventemitter.EventEmitter), _temp), (_descriptor = (0, _applyDecoratedDescriptor2.default)(_class2.prototype, "id", [_dec2], {
  configurable: true,
  enumerable: true,
  writable: true,
  initializer: null
}), _descriptor2 = (0, _applyDecoratedDescriptor2.default)(_class2.prototype, "iconService", [_dec3], {
  configurable: true,
  enumerable: true,
  writable: true,
  initializer: null
}), _descriptor3 = (0, _applyDecoratedDescriptor2.default)(_class2.prototype, "fontService", [_dec4], {
  configurable: true,
  enumerable: true,
  writable: true,
  initializer: null
}), _descriptor4 = (0, _applyDecoratedDescriptor2.default)(_class2.prototype, "controlService", [_dec5], {
  configurable: true,
  enumerable: true,
  writable: true,
  initializer: null
}), _descriptor5 = (0, _applyDecoratedDescriptor2.default)(_class2.prototype, "logger", [_dec6], {
  configurable: true,
  enumerable: true,
  writable: true,
  initializer: null
}), _descriptor6 = (0, _applyDecoratedDescriptor2.default)(_class2.prototype, "configService", [_dec7], {
  configurable: true,
  enumerable: true,
  writable: true,
  initializer: null
}), _descriptor7 = (0, _applyDecoratedDescriptor2.default)(_class2.prototype, "map", [_dec8], {
  configurable: true,
  enumerable: true,
  writable: true,
  initializer: null
}), _descriptor8 = (0, _applyDecoratedDescriptor2.default)(_class2.prototype, "coordinateSystemService", [_dec9], {
  configurable: true,
  enumerable: true,
  writable: true,
  initializer: null
}), _descriptor9 = (0, _applyDecoratedDescriptor2.default)(_class2.prototype, "rendererService", [_dec10], {
  configurable: true,
  enumerable: true,
  writable: true,
  initializer: null
}), _descriptor10 = (0, _applyDecoratedDescriptor2.default)(_class2.prototype, "layerService", [_dec11], {
  configurable: true,
  enumerable: true,
  writable: true,
  initializer: null
}), _descriptor11 = (0, _applyDecoratedDescriptor2.default)(_class2.prototype, "cameraService", [_dec12], {
  configurable: true,
  enumerable: true,
  writable: true,
  initializer: null
}), _descriptor12 = (0, _applyDecoratedDescriptor2.default)(_class2.prototype, "interactionService", [_dec13], {
  configurable: true,
  enumerable: true,
  writable: true,
  initializer: null
}), _descriptor13 = (0, _applyDecoratedDescriptor2.default)(_class2.prototype, "pickingService", [_dec14], {
  configurable: true,
  enumerable: true,
  writable: true,
  initializer: null
}), _descriptor14 = (0, _applyDecoratedDescriptor2.default)(_class2.prototype, "shaderModuleService", [_dec15], {
  configurable: true,
  enumerable: true,
  writable: true,
  initializer: null
}), _descriptor15 = (0, _applyDecoratedDescriptor2.default)(_class2.prototype, "markerService", [_dec16], {
  configurable: true,
  enumerable: true,
  writable: true,
  initializer: null
}), _descriptor16 = (0, _applyDecoratedDescriptor2.default)(_class2.prototype, "popupService", [_dec17], {
  configurable: true,
  enumerable: true,
  writable: true,
  initializer: null
})), _class2)) || _class);
exports.default = Scene;
