// @antv/l7-layers/lib/utils/blend.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.BlendTypes = void 0;

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _l7Core = require("@antv/l7-core");

var _BlendTypes;

var BlendTypes = (_BlendTypes = {}, (0, _defineProperty2.default)(_BlendTypes, _l7Core.BlendType.additive, {
  enable: true,
  func: {
    srcRGB: _l7Core.gl.ONE,
    dstRGB: _l7Core.gl.ONE,
    srcAlpha: 1,
    dstAlpha: 1
  }
}), (0, _defineProperty2.default)(_BlendTypes, _l7Core.BlendType.none, {
  enable: false
}), (0, _defineProperty2.default)(_BlendTypes, _l7Core.BlendType.normal, {
  enable: true,
  func: {
    srcRGB: _l7Core.gl.SRC_ALPHA,
    dstRGB: _l7Core.gl.ONE_MINUS_SRC_ALPHA,
    srcAlpha: 1,
    dstAlpha: 1
  }
}), (0, _defineProperty2.default)(_BlendTypes, _l7Core.BlendType.subtractive, {
  enable: true,
  func: {
    srcRGB: _l7Core.gl.ONE,
    dstRGB: _l7Core.gl.ONE,
    srcAlpha: _l7Core.gl.ZERO,
    dstAlpha: _l7Core.gl.ONE_MINUS_SRC_COLOR
  },
  equation: {
    rgb: _l7Core.gl.FUNC_SUBTRACT,
    alpha: _l7Core.gl.FUNC_SUBTRACT
  }
}), (0, _defineProperty2.default)(_BlendTypes, _l7Core.BlendType.max, {
  enable: true,
  func: {
    srcRGB: _l7Core.gl.ONE,
    dstRGB: _l7Core.gl.ONE
  },
  equation: {
    rgb: _l7Core.gl.MAX_EXT
  }
}), (0, _defineProperty2.default)(_BlendTypes, _l7Core.BlendType.min, {
  enable: true,
  func: {
    srcRGB: _l7Core.gl.ONE,
    dstRGB: _l7Core.gl.ONE
  },
  equation: {
    rgb: _l7Core.gl.MIN_EXT
  }
}), _BlendTypes);
exports.BlendTypes = BlendTypes;
