// @antv/l7-map/lib/handler/touch/touch_zoom.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _get2 = _interopRequireDefault(require("@babel/runtime/helpers/get"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _two_touch = _interopRequireDefault(require("./two_touch"));

function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = (0, _getPrototypeOf2.default)(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = (0, _getPrototypeOf2.default)(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return (0, _possibleConstructorReturn2.default)(this, result); }; }

function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Boolean.prototype.valueOf.call(Reflect.construct(Boolean, [], function () {})); return true; } catch (e) { return false; } }

var ZOOM_THRESHOLD = 0.1;

function getZoomDelta(distance, lastDistance) {
  return Math.log(distance / lastDistance) / Math.LN2;
}

var TouchZoomHandler = function (_TwoTouchHandler) {
  (0, _inherits2.default)(TouchZoomHandler, _TwoTouchHandler);

  var _super = _createSuper(TouchZoomHandler);

  function TouchZoomHandler() {
    var _this;

    (0, _classCallCheck2.default)(this, TouchZoomHandler);

    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    _this = _super.call.apply(_super, [this].concat(args));
    _this.distance = void 0;
    _this.startDistance = void 0;
    return _this;
  }

  (0, _createClass2.default)(TouchZoomHandler, [{
    key: "reset",
    value: function reset() {
      (0, _get2.default)((0, _getPrototypeOf2.default)(TouchZoomHandler.prototype), "reset", this).call(this);
      delete this.distance;
      delete this.startDistance;
    }
  }, {
    key: "start",
    value: function start(points) {
      this.startDistance = this.distance = points[0].dist(points[1]);
    }
  }, {
    key: "move",
    value: function move(points, pinchAround) {
      var lastDistance = this.distance;
      this.distance = points[0].dist(points[1]);

      if (!this.active && Math.abs(getZoomDelta(this.distance, this.startDistance)) < ZOOM_THRESHOLD) {
        return;
      }

      this.active = true;
      return {
        zoomDelta: getZoomDelta(this.distance, lastDistance),
        pinchAround: pinchAround
      };
    }
  }]);
  return TouchZoomHandler;
}(_two_touch.default);

exports.default = TouchZoomHandler;
