// @antv/l7-react/es/component/Legend/color.js
import * as React from 'react';
export var ColorComponent = /*#__PURE__*/React.memo(function (props) {
  // const { className, style, title } = props;
  var defaultItems = [{
    title: '1',
    color: 'rgb(239,243,255)'
  }, {
    title: '10',
    color: 'rgb(198,219,239)'
  }, {
    title: '30',
    color: 'rgb(158,202,225)'
  }, {
    title: '50',
    color: 'rgb(107,174,214)'
  }, {
    title: '60',
    color: 'rgb(49,130,189)'
  }, {
    title: '100',
    color: 'rgb(8,81,156)'
  }];
  var _props$items = props.items,
      items = _props$items === void 0 ? defaultItems : _props$items;
  return /*#__PURE__*/React.createElement("div", null, /*#__PURE__*/React.createElement("div", null, items.map(function (c, i) {
    return /*#__PURE__*/React.createElement("div", {
      key: i,
      style: {
        background: c.color,
        height: '100%',
        display: 'inline-block',
        cursor: 'pointer',
        width: '' + (100.0 - items.length) / items.length + '%',
        marginRight: '1%',
        padding: 5
      }
    });
  })), /*#__PURE__*/React.createElement("div", null, items.map(function (c, i) {
    return /*#__PURE__*/React.createElement("div", {
      key: i,
      style: {
        background: '#fff',
        height: '100%',
        display: 'inline-block',
        textAlign: 'left',
        cursor: 'pointer',
        width: '' + (100.0 - items.length) / items.length + '%',
        marginRight: '1%'
      }
    }, c.title);
  })));
});
