// gg-editor/cjs/helpers/track.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

require("core-js/modules/es.array.concat");

require("core-js/modules/es.object.define-property");

require("core-js/modules/web.timers");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = void 0;

var _objectSpread2 = _interopRequireDefault(require("@babel/runtime/helpers/objectSpread"));

var _Global = _interopRequireDefault(require("../common/Global"));

var _utils = require("../utils");

var BASE_URL = 'http://gm.mmstat.com/fsp.1.1';

var track = function track(options) {
  var trackable = _Global["default"].get('trackable');

  var version = _Global["default"].get('version');

  if (!trackable) {
    return;
  }

  var _window = window,
      location = _window.location,
      navigator = _window.navigator;
  var image = new Image();
  var params = (0, _utils.toQueryString)((0, _objectSpread2["default"])({
    pid: 'ggeditor',
    code: '11',
    msg: 'syslog',
    page: "".concat(location.protocol, "//").concat(location.host).concat(location.pathname),
    hash: location.hash,
    ua: navigator.userAgent,
    rel: version
  }, options));
  image.src = "".concat(BASE_URL, "?").concat(params);
};

var _default = function _default(options) {
  setTimeout(function () {
    track(options);
  }, 1000);
};

exports["default"] = _default;
