// @antv/l7-map/es/geo/mercator.js
import _classCallCheck from "@babel/runtime/helpers/classCallCheck";
import _createClass from "@babel/runtime/helpers/createClass";
import LngLat, { earthRadius } from '../geo/lng_lat';
var earthCircumfrence = 2 * Math.PI * earthRadius;

function circumferenceAtLatitude(latitude) {
  return earthCircumfrence * Math.cos(latitude * Math.PI / 180);
}

export function mercatorXfromLng(lng) {
  return (180 + lng) / 360;
}
export function mercatorYfromLat(lat) {
  return (180 - 180 / Math.PI * Math.log(Math.tan(Math.PI / 4 + lat * Math.PI / 360))) / 360;
}
export function mercatorZfromAltitude(altitude, lat) {
  return altitude / circumferenceAtLatitude(lat);
}
export function lngFromMercatorX(x) {
  return x * 360 - 180;
}
export function latFromMercatorY(y) {
  var y2 = 180 - y * 360;
  return 360 / Math.PI * Math.atan(Math.exp(y2 * Math.PI / 180)) - 90;
}
export function altitudeFromMercatorZ(z, y) {
  return z * circumferenceAtLatitude(latFromMercatorY(y));
}
export function mercatorScale(lat) {
  return 1 / Math.cos(lat * Math.PI / 180);
}

var MercatorCoordinate = function () {
  function MercatorCoordinate(x, y) {
    var z = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : 0;

    _classCallCheck(this, MercatorCoordinate);

    this.x = void 0;
    this.y = void 0;
    this.z = void 0;
    this.x = +x;
    this.y = +y;
    this.z = +z;
  }

  _createClass(MercatorCoordinate, [{
    key: "toLngLat",
    value: function toLngLat() {
      return new LngLat(lngFromMercatorX(this.x), latFromMercatorY(this.y));
    }
  }, {
    key: "toAltitude",
    value: function toAltitude() {
      return altitudeFromMercatorZ(this.z, this.y);
    }
  }, {
    key: "meterInMercatorCoordinateUnits",
    value: function meterInMercatorCoordinateUnits() {
      return 1 / earthCircumfrence * mercatorScale(latFromMercatorY(this.y));
    }
  }], [{
    key: "fromLngLat",
    value: function fromLngLat(lngLatLike) {
      var altitude = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : 0;
      var lngLat = LngLat.convert(lngLatLike);
      return new MercatorCoordinate(mercatorXfromLng(lngLat.lng), mercatorYfromLat(lngLat.lat), mercatorZfromAltitude(altitude, lngLat.lat));
    }
  }]);

  return MercatorCoordinate;
}();

export { MercatorCoordinate as default };
