// @antv/l7-core/lib/services/coordinate/CoordinateSystemService.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _initializerDefineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/initializerDefineProperty"));

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _applyDecoratedDescriptor2 = _interopRequireDefault(require("@babel/runtime/helpers/applyDecoratedDescriptor"));

var _initializerWarningHelper2 = _interopRequireDefault(require("@babel/runtime/helpers/initializerWarningHelper"));

var _glMatrix = require("gl-matrix");

var _inversify = require("inversify");

var _types = require("../../types");

var _project = require("../../utils/project");

var _ICoordinateSystemService = require("./ICoordinateSystemService");

var _dec, _dec2, _class, _class2, _descriptor, _temp;

var VECTOR_TO_POINT_MATRIX = [1, 0, 0, 0, 0, 1, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0];
var CoordinateSystemService = (_dec = (0, _inversify.injectable)(), _dec2 = (0, _inversify.inject)(_types.TYPES.ICameraService), _dec(_class = (_class2 = (_temp = function () {
  function CoordinateSystemService() {
    (0, _classCallCheck2.default)(this, CoordinateSystemService);
    this.needRefresh = true;
    (0, _initializerDefineProperty2.default)(this, "cameraService", _descriptor, this);
    this.coordinateSystem = void 0;
    this.viewportCenter = void 0;
    this.viewportCenterProjection = void 0;
    this.pixelsPerDegree = void 0;
    this.pixelsPerDegree2 = void 0;
    this.pixelsPerMeter = void 0;
  }

  (0, _createClass2.default)(CoordinateSystemService, [{
    key: "refresh",
    value: function refresh() {
      var zoom = this.cameraService.getZoom();
      var zoomScale = this.cameraService.getZoomScale();
      var center = this.cameraService.getCenter();

      var _getDistanceScales = (0, _project.getDistanceScales)({
        latitude: center[1],
        zoom: zoom
      }),
          pixelsPerMeter = _getDistanceScales.pixelsPerMeter,
          pixelsPerDegree = _getDistanceScales.pixelsPerDegree;

      this.viewportCenter = center;
      this.viewportCenterProjection = [0, 0, 0, 0];
      this.pixelsPerMeter = pixelsPerMeter;
      this.pixelsPerDegree = pixelsPerDegree;
      this.pixelsPerDegree2 = [0, 0, 0];

      if (this.coordinateSystem === _ICoordinateSystemService.CoordinateSystem.LNGLAT || this.coordinateSystem === _ICoordinateSystemService.CoordinateSystem.P20) {
        this.cameraService.setViewProjectionMatrix(undefined);
      } else if (this.coordinateSystem === _ICoordinateSystemService.CoordinateSystem.LNGLAT_OFFSET) {
        this.calculateLnglatOffset(center, zoom);
      } else if (this.coordinateSystem === _ICoordinateSystemService.CoordinateSystem.P20_OFFSET) {
        this.calculateLnglatOffset(center, zoom, zoomScale, true);
      }

      this.needRefresh = false;
    }
  }, {
    key: "getCoordinateSystem",
    value: function getCoordinateSystem() {
      return this.coordinateSystem;
    }
  }, {
    key: "setCoordinateSystem",
    value: function setCoordinateSystem(coordinateSystem) {
      this.coordinateSystem = coordinateSystem;
    }
  }, {
    key: "getViewportCenter",
    value: function getViewportCenter() {
      return this.viewportCenter;
    }
  }, {
    key: "getViewportCenterProjection",
    value: function getViewportCenterProjection() {
      return this.viewportCenterProjection;
    }
  }, {
    key: "getPixelsPerDegree",
    value: function getPixelsPerDegree() {
      return this.pixelsPerDegree;
    }
  }, {
    key: "getPixelsPerDegree2",
    value: function getPixelsPerDegree2() {
      return this.pixelsPerDegree2;
    }
  }, {
    key: "getPixelsPerMeter",
    value: function getPixelsPerMeter() {
      return this.pixelsPerMeter;
    }
  }, {
    key: "calculateLnglatOffset",
    value: function calculateLnglatOffset(center, zoom, scale, flipY) {
      var _getDistanceScales2 = (0, _project.getDistanceScales)({
        latitude: center[1],
        zoom: zoom,
        scale: scale,
        flipY: flipY,
        highPrecision: true
      }),
          ppm = _getDistanceScales2.pixelsPerMeter,
          ppd = _getDistanceScales2.pixelsPerDegree,
          pixelsPerDegree2 = _getDistanceScales2.pixelsPerDegree2;

      var viewMatrix = this.cameraService.getViewMatrix();
      var projectionMatrix = this.cameraService.getProjectionMatrix();

      var viewProjectionMatrix = _glMatrix.mat4.multiply([], projectionMatrix, viewMatrix);

      var positionPixels = this.cameraService.projectFlat([Math.fround(center[0]), Math.fround(center[1])], Math.pow(2, zoom));
      this.viewportCenterProjection = _glMatrix.vec4.transformMat4([], [positionPixels[0], positionPixels[1], 0.0, 1.0], viewProjectionMatrix);
      viewMatrix = this.cameraService.getViewMatrixUncentered() || viewMatrix;
      viewProjectionMatrix = _glMatrix.mat4.multiply([], projectionMatrix, viewMatrix);
      viewProjectionMatrix = _glMatrix.mat4.multiply([], viewProjectionMatrix, VECTOR_TO_POINT_MATRIX);
      this.cameraService.setViewProjectionMatrix(viewProjectionMatrix);
      this.pixelsPerMeter = ppm;
      this.pixelsPerDegree = ppd;
      this.pixelsPerDegree2 = pixelsPerDegree2;
    }
  }]);
  return CoordinateSystemService;
}(), _temp), (_descriptor = (0, _applyDecoratedDescriptor2.default)(_class2.prototype, "cameraService", [_dec2], {
  configurable: true,
  enumerable: true,
  writable: true,
  initializer: null
})), _class2)) || _class);
exports.default = CoordinateSystemService;
