// @antv/l7-core/lib/services/coordinate/ICoordinateSystemService.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.CoordinateUniform = exports.CoordinateSystem = void 0;
var CoordinateSystem;
exports.CoordinateSystem = CoordinateSystem;

(function (CoordinateSystem) {
  CoordinateSystem[CoordinateSystem["LNGLAT"] = 1] = "LNGLAT";
  CoordinateSystem[CoordinateSystem["LNGLAT_OFFSET"] = 2] = "LNGLAT_OFFSET";
  CoordinateSystem[CoordinateSystem["VECTOR_TILE"] = 3] = "VECTOR_TILE";
  CoordinateSystem[CoordinateSystem["IDENTITY"] = 4] = "IDENTITY";
  CoordinateSystem[CoordinateSystem["P20"] = 5] = "P20";
  CoordinateSystem[CoordinateSystem["P20_OFFSET"] = 6] = "P20_OFFSET";
  CoordinateSystem[CoordinateSystem["METER_OFFSET"] = 7] = "METER_OFFSET";
})(CoordinateSystem || (exports.CoordinateSystem = CoordinateSystem = {}));

var CoordinateUniform = {
  CoordinateSystem: 'u_CoordinateSystem',
  ViewportCenter: 'u_ViewportCenter',
  ViewportCenterProjection: 'u_ViewportCenterProjection',
  PixelsPerDegree: 'u_PixelsPerDegree',
  PixelsPerDegree2: 'u_PixelsPerDegree2',
  PixelsPerMeter: 'u_PixelsPerMeter'
};
exports.CoordinateUniform = CoordinateUniform;
