// @antv/l7-layers/lib/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "CityBuildingLayer", {
  enumerable: true,
  get: function get() {
    return _building.default;
  }
});
Object.defineProperty(exports, "BaseLayer", {
  enumerable: true,
  get: function get() {
    return _BaseLayer.default;
  }
});
Object.defineProperty(exports, "HeatmapLayer", {
  enumerable: true,
  get: function get() {
    return _heatmap.default;
  }
});
Object.defineProperty(exports, "ImageLayer", {
  enumerable: true,
  get: function get() {
    return _image.default;
  }
});
Object.defineProperty(exports, "LineLayer", {
  enumerable: true,
  get: function get() {
    return _index.default;
  }
});
Object.defineProperty(exports, "PointLayer", {
  enumerable: true,
  get: function get() {
    return _point.default;
  }
});
Object.defineProperty(exports, "PolygonLayer", {
  enumerable: true,
  get: function get() {
    return _polygon.default;
  }
});
Object.defineProperty(exports, "RasterLayer", {
  enumerable: true,
  get: function get() {
    return _raster.default;
  }
});

var _l7Core = require("@antv/l7-core");

var _building = _interopRequireDefault(require("./citybuliding/building"));

var _BaseLayer = _interopRequireDefault(require("./core/BaseLayer"));

require("./glsl.d");

var _heatmap = _interopRequireDefault(require("./heatmap"));

var _image = _interopRequireDefault(require("./image"));

var _index = _interopRequireDefault(require("./line/index"));

var _point = _interopRequireDefault(require("./point"));

var _polygon = _interopRequireDefault(require("./polygon"));

var _raster = _interopRequireDefault(require("./raster"));

var _DataMappingPlugin = _interopRequireDefault(require("./plugins/DataMappingPlugin"));

var _DataSourcePlugin = _interopRequireDefault(require("./plugins/DataSourcePlugin"));

var _FeatureScalePlugin = _interopRequireDefault(require("./plugins/FeatureScalePlugin"));

var _LayerAnimateStylePlugin = _interopRequireDefault(require("./plugins/LayerAnimateStylePlugin"));

var _LayerModelPlugin = _interopRequireDefault(require("./plugins/LayerModelPlugin"));

var _LayerStylePlugin = _interopRequireDefault(require("./plugins/LayerStylePlugin"));

var _LightingPlugin = _interopRequireDefault(require("./plugins/LightingPlugin"));

var _MultiPassRendererPlugin = _interopRequireDefault(require("./plugins/MultiPassRendererPlugin"));

var _PixelPickingPlugin = _interopRequireDefault(require("./plugins/PixelPickingPlugin"));

var _RegisterStyleAttributePlugin = _interopRequireDefault(require("./plugins/RegisterStyleAttributePlugin"));

var _ShaderUniformPlugin = _interopRequireDefault(require("./plugins/ShaderUniformPlugin"));

var _UpdateModelPlugin = _interopRequireDefault(require("./plugins/UpdateModelPlugin"));

var _UpdateStyleAttributePlugin = _interopRequireDefault(require("./plugins/UpdateStyleAttributePlugin"));

_l7Core.container.bind(_l7Core.TYPES.ILayerPlugin).to(_DataSourcePlugin.default).inRequestScope();

_l7Core.container.bind(_l7Core.TYPES.ILayerPlugin).to(_RegisterStyleAttributePlugin.default).inRequestScope();

_l7Core.container.bind(_l7Core.TYPES.ILayerPlugin).to(_FeatureScalePlugin.default).inRequestScope();

_l7Core.container.bind(_l7Core.TYPES.ILayerPlugin).to(_DataMappingPlugin.default).inRequestScope();

_l7Core.container.bind(_l7Core.TYPES.ILayerPlugin).to(_LayerStylePlugin.default).inRequestScope();

_l7Core.container.bind(_l7Core.TYPES.ILayerPlugin).to(_UpdateStyleAttributePlugin.default).inRequestScope();

_l7Core.container.bind(_l7Core.TYPES.ILayerPlugin).to(_UpdateModelPlugin.default).inRequestScope();

_l7Core.container.bind(_l7Core.TYPES.ILayerPlugin).to(_MultiPassRendererPlugin.default).inRequestScope();

_l7Core.container.bind(_l7Core.TYPES.ILayerPlugin).to(_ShaderUniformPlugin.default).inRequestScope();

_l7Core.container.bind(_l7Core.TYPES.ILayerPlugin).to(_LayerAnimateStylePlugin.default).inRequestScope();

_l7Core.container.bind(_l7Core.TYPES.ILayerPlugin).to(_LightingPlugin.default).inRequestScope();

_l7Core.container.bind(_l7Core.TYPES.ILayerPlugin).to(_PixelPickingPlugin.default).inRequestScope();

_l7Core.container.bind(_l7Core.TYPES.ILayerPlugin).to(_LayerModelPlugin.default).inRequestScope();
