// @antv/l7-core/lib/services/log/LogService.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _inversify = require("inversify");

var _probe = require("probe.gl");

var _dec, _class;

var Logger = new _probe.Log({
  id: 'L7'
}).enable(false);
Logger.priority = 5;
var LogService = (_dec = (0, _inversify.injectable)(), _dec(_class = function () {
  function LogService() {
    (0, _classCallCheck2.default)(this, LogService);
  }

  (0, _createClass2.default)(LogService, [{
    key: "error",
    value: function error(message) {
      Logger.error(message)();
    }
  }, {
    key: "warn",
    value: function warn(message) {
      Logger.probe(1, message)();
    }
  }, {
    key: "info",
    value: function info(message) {
      Logger.info(3, message)();
    }
  }, {
    key: "debug",
    value: function debug(message) {
      Logger.probe(4, message)();
    }
  }]);
  return LogService;
}()) || _class);
exports.default = LogService;
