// @antv/l7-map/lib/handler/blockable_map_event.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _events = require("./events");

var BlockableMapEventHandler = function () {
  function BlockableMapEventHandler(map) {
    (0, _classCallCheck2.default)(this, BlockableMapEventHandler);
    this.map = void 0;
    this.delayContextMenu = void 0;
    this.contextMenuEvent = void 0;
    this.map = map;
  }

  (0, _createClass2.default)(BlockableMapEventHandler, [{
    key: "reset",
    value: function reset() {
      this.delayContextMenu = false;
      delete this.contextMenuEvent;
    }
  }, {
    key: "mousemove",
    value: function mousemove(e) {
      this.map.emit(e.type, new _events.MapMouseEvent(e.type, this.map, e));
    }
  }, {
    key: "mousedown",
    value: function mousedown() {
      this.delayContextMenu = true;
    }
  }, {
    key: "mouseup",
    value: function mouseup() {
      this.delayContextMenu = false;

      if (this.contextMenuEvent) {
        this.map.emit('contextmenu', new _events.MapMouseEvent('contextmenu', this.map, this.contextMenuEvent));
        delete this.contextMenuEvent;
      }
    }
  }, {
    key: "contextmenu",
    value: function contextmenu(e) {
      if (this.delayContextMenu) {
        this.contextMenuEvent = e;
      } else {
        this.map.emit(e.type, new _events.MapMouseEvent(e.type, this.map, e));
      }

      if (this.map.listeners('contextmenu')) {
        e.preventDefault();
      }
    }
  }, {
    key: "isEnabled",
    value: function isEnabled() {
      return true;
    }
  }, {
    key: "isActive",
    value: function isActive() {
      return false;
    }
  }, {
    key: "enable",
    value: function enable() {
      return true;
    }
  }, {
    key: "disable",
    value: function disable() {
      return false;
    }
  }]);
  return BlockableMapEventHandler;
}();

exports.default = BlockableMapEventHandler;
