// @antv/l7-map/es/handler/touch/touch_pan.js
import _classCallCheck from "@babel/runtime/helpers/classCallCheck";
import _createClass from "@babel/runtime/helpers/createClass";
import Point from '../../geo/point';
import { indexTouches } from '../handler_util';

var TouchPanHandler = function () {
  function TouchPanHandler(options) {
    _classCallCheck(this, TouchPanHandler);

    this.enabled = void 0;
    this.active = void 0;
    this.touches = void 0;
    this.minTouches = void 0;
    this.clickTolerance = void 0;
    this.sum = void 0;
    this.minTouches = 1;
    this.clickTolerance = options.clickTolerance || 1;
    this.reset();
  }

  _createClass(TouchPanHandler, [{
    key: "reset",
    value: function reset() {
      this.active = false;
      this.touches = {};
      this.sum = new Point(0, 0);
    }
  }, {
    key: "touchstart",
    value: function touchstart(e, points, mapTouches) {
      return this.calculateTransform(e, points, mapTouches);
    }
  }, {
    key: "touchmove",
    value: function touchmove(e, points, mapTouches) {
      if (!this.active) {
        return;
      }

      e.preventDefault();
      return this.calculateTransform(e, points, mapTouches);
    }
  }, {
    key: "touchend",
    value: function touchend(e, points, mapTouches) {
      this.calculateTransform(e, points, mapTouches);

      if (this.active && mapTouches.length < this.minTouches) {
        this.reset();
      }
    }
  }, {
    key: "touchcancel",
    value: function touchcancel() {
      this.reset();
    }
  }, {
    key: "enable",
    value: function enable() {
      this.enabled = true;
    }
  }, {
    key: "disable",
    value: function disable() {
      this.enabled = false;
      this.reset();
    }
  }, {
    key: "isEnabled",
    value: function isEnabled() {
      return this.enabled;
    }
  }, {
    key: "isActive",
    value: function isActive() {
      return this.active;
    }
  }, {
    key: "calculateTransform",
    value: function calculateTransform(e, points, mapTouches) {
      if (mapTouches.length > 0) {
        this.active = true;
      }

      var touches = indexTouches(mapTouches, points);
      var touchPointSum = new Point(0, 0);
      var touchDeltaSum = new Point(0, 0);
      var touchDeltaCount = 0;

      for (var identifier in touches) {
        if (touches[identifier]) {
          var point = touches[identifier];
          var prevPoint = this.touches[identifier];

          if (prevPoint) {
            touchPointSum._add(point);

            touchDeltaSum._add(point.sub(prevPoint));

            touchDeltaCount++;
            touches[identifier] = point;
          }
        }
      }

      this.touches = touches;

      if (touchDeltaCount < this.minTouches || !touchDeltaSum.mag()) {
        return;
      }

      var panDelta = touchDeltaSum.div(touchDeltaCount);

      this.sum._add(panDelta);

      if (this.sum.mag() < this.clickTolerance) {
        return;
      }

      var around = touchPointSum.div(touchDeltaCount);
      return {
        around: around,
        panDelta: panDelta
      };
    }
  }]);

  return TouchPanHandler;
}();

export { TouchPanHandler as default };
