// @antv/l7-map/es/handler/touch/touch_rotate.js
import _classCallCheck from "@babel/runtime/helpers/classCallCheck";
import _createClass from "@babel/runtime/helpers/createClass";
import _get from "@babel/runtime/helpers/get";
import _inherits from "@babel/runtime/helpers/inherits";
import _possibleConstructorReturn from "@babel/runtime/helpers/possibleConstructorReturn";
import _getPrototypeOf from "@babel/runtime/helpers/getPrototypeOf";

function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = _getPrototypeOf(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = _getPrototypeOf(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return _possibleConstructorReturn(this, result); }; }

function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Boolean.prototype.valueOf.call(Reflect.construct(Boolean, [], function () {})); return true; } catch (e) { return false; } }

import TwoTouchHandler from './two_touch';
var ROTATION_THRESHOLD = 25;

function getBearingDelta(a, b) {
  return a.angleWith(b) * 180 / Math.PI;
}

var TouchRotateHandler = function (_TwoTouchHandler) {
  _inherits(TouchRotateHandler, _TwoTouchHandler);

  var _super = _createSuper(TouchRotateHandler);

  function TouchRotateHandler() {
    var _this;

    _classCallCheck(this, TouchRotateHandler);

    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    _this = _super.call.apply(_super, [this].concat(args));
    _this.minDiameter = void 0;
    return _this;
  }

  _createClass(TouchRotateHandler, [{
    key: "reset",
    value: function reset() {
      _get(_getPrototypeOf(TouchRotateHandler.prototype), "reset", this).call(this);

      delete this.minDiameter;
      delete this.startVector;
      delete this.vector;
    }
  }, {
    key: "start",
    value: function start(points) {
      this.startVector = this.vector = points[0].sub(points[1]);
      this.minDiameter = points[0].dist(points[1]);
    }
  }, {
    key: "move",
    value: function move(points, pinchAround) {
      var lastVector = this.vector;
      this.vector = points[0].sub(points[1]);

      if (!this.active && this.isBelowThreshold(this.vector)) {
        return;
      }

      this.active = true;
      return {
        bearingDelta: getBearingDelta(this.vector, lastVector),
        pinchAround: pinchAround
      };
    }
  }, {
    key: "isBelowThreshold",
    value: function isBelowThreshold(vector) {
      this.minDiameter = Math.min(this.minDiameter, vector.mag());
      var circumference = Math.PI * this.minDiameter;
      var threshold = ROTATION_THRESHOLD / circumference * 360;
      var bearingDeltaSinceStart = getBearingDelta(vector, this.startVector);
      return Math.abs(bearingDeltaSinceStart) < threshold;
    }
  }]);

  return TouchRotateHandler;
}(TwoTouchHandler);

export { TouchRotateHandler as default };
