// @antv/l7-core/lib/services/layer/IStyleAttributeService.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.AttributeType = exports.StyleScaleType = exports.ScaleTypes = void 0;
var ScaleTypes;
exports.ScaleTypes = ScaleTypes;

(function (ScaleTypes) {
  ScaleTypes["LINEAR"] = "linear";
  ScaleTypes["POWER"] = "power";
  ScaleTypes["LOG"] = "log";
  ScaleTypes["IDENTITY"] = "identity";
  ScaleTypes["TIME"] = "time";
  ScaleTypes["QUANTILE"] = "quantile";
  ScaleTypes["QUANTIZE"] = "quantize";
  ScaleTypes["THRESHOLD"] = "threshold";
  ScaleTypes["CAT"] = "cat";
})(ScaleTypes || (exports.ScaleTypes = ScaleTypes = {}));

var StyleScaleType;
exports.StyleScaleType = StyleScaleType;

(function (StyleScaleType) {
  StyleScaleType["CONSTANT"] = "constant";
  StyleScaleType["VARIABLE"] = "variable";
})(StyleScaleType || (exports.StyleScaleType = StyleScaleType = {}));

var AttributeType;
exports.AttributeType = AttributeType;

(function (AttributeType) {
  AttributeType[AttributeType["Attribute"] = 0] = "Attribute";
  AttributeType[AttributeType["InstancedAttribute"] = 1] = "InstancedAttribute";
  AttributeType[AttributeType["Uniform"] = 2] = "Uniform";
})(AttributeType || (exports.AttributeType = AttributeType = {}));
