// @antv/l7-core/lib/services/layer/StyleAttribute.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _toConsumableArray2 = _interopRequireDefault(require("@babel/runtime/helpers/toConsumableArray"));

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _isNil2 = _interopRequireDefault(require("lodash/isNil"));

var StyleAttribute = function () {
  function StyleAttribute(options) {
    var _this = this;

    (0, _classCallCheck2.default)(this, StyleAttribute);
    this.name = void 0;
    this.type = void 0;
    this.scale = void 0;
    this.descriptor = void 0;
    this.featureBufferLayout = [];
    this.needRescale = false;
    this.needRemapping = false;
    this.needRegenerateVertices = false;
    this.featureRange = {
      startIndex: 0,
      endIndex: Infinity
    };
    this.vertexAttribute = void 0;

    this.defaultCallback = function (params) {
      if (params.length === 0) {
        var _this$scale;

        return ((_this$scale = _this.scale) === null || _this$scale === void 0 ? void 0 : _this$scale.defaultValues) || [];
      }

      return params.map(function (param, idx) {
        var _this$scale2;

        var scaleFunc = (_this$scale2 = _this.scale) === null || _this$scale2 === void 0 ? void 0 : _this$scale2.scalers[idx].func;
        var value = scaleFunc(param);
        return value;
      });
    };

    this.setProps(options);
  }

  (0, _createClass2.default)(StyleAttribute, [{
    key: "setProps",
    value: function setProps(options) {
      Object.assign(this, options);
    }
  }, {
    key: "mapping",
    value: function mapping(params) {
      var _this$scale3;

      if ((_this$scale3 = this.scale) !== null && _this$scale3 !== void 0 && _this$scale3.callback) {
        var _this$scale4;

        var ret = (_this$scale4 = this.scale) === null || _this$scale4 === void 0 ? void 0 : _this$scale4.callback.apply(_this$scale4, (0, _toConsumableArray2.default)(params));

        if (!(0, _isNil2.default)(ret)) {
          return [ret];
        }
      }

      return this.defaultCallback(params);
    }
  }, {
    key: "resetDescriptor",
    value: function resetDescriptor() {
      if (this.descriptor) {
        this.descriptor.buffer.data = [];
      }
    }
  }]);
  return StyleAttribute;
}();

exports.default = StyleAttribute;
