// @antv/l7-map/es/handler/shim/dblclick_zoom.js
import _classCallCheck from "@babel/runtime/helpers/classCallCheck";
import _createClass from "@babel/runtime/helpers/createClass";

var DoubleClickZoomHandler = function () {
  function DoubleClickZoomHandler(clickZoom, TapZoom) {
    _classCallCheck(this, DoubleClickZoomHandler);

    this.clickZoom = void 0;
    this.tapZoom = void 0;
    this.clickZoom = clickZoom;
    this.tapZoom = TapZoom;
  }

  _createClass(DoubleClickZoomHandler, [{
    key: "enable",
    value: function enable() {
      this.clickZoom.enable();
      this.tapZoom.enable();
    }
  }, {
    key: "disable",
    value: function disable() {
      this.clickZoom.disable();
      this.tapZoom.disable();
    }
  }, {
    key: "isEnabled",
    value: function isEnabled() {
      return this.clickZoom.isEnabled() && this.tapZoom.isEnabled();
    }
  }, {
    key: "isActive",
    value: function isActive() {
      return this.clickZoom.isActive() || this.tapZoom.isActive();
    }
  }]);

  return DoubleClickZoomHandler;
}();

export { DoubleClickZoomHandler as default };
