// @antv/l7-map/es/handler/shim/drag_pan.js
import _classCallCheck from "@babel/runtime/helpers/classCallCheck";
import _createClass from "@babel/runtime/helpers/createClass";

var DragPanHandler = function () {
  function DragPanHandler(el, mousePan, touchPan) {
    _classCallCheck(this, DragPanHandler);

    this.inertiaOptions = void 0;
    this.el = void 0;
    this.mousePan = void 0;
    this.touchPan = void 0;
    this.el = el;
    this.mousePan = mousePan;
    this.touchPan = touchPan;
  }

  _createClass(DragPanHandler, [{
    key: "enable",
    value: function enable(options) {
      this.inertiaOptions = options || {};
      this.mousePan.enable();
      this.touchPan.enable();
      this.el.classList.add('l7-touch-drag-pan');
    }
  }, {
    key: "disable",
    value: function disable() {
      this.mousePan.disable();
      this.touchPan.disable();
      this.el.classList.remove('l7-touch-drag-pan');
    }
  }, {
    key: "isEnabled",
    value: function isEnabled() {
      return this.mousePan.isEnabled() && this.touchPan.isEnabled();
    }
  }, {
    key: "isActive",
    value: function isActive() {
      return this.mousePan.isActive() || this.touchPan.isActive();
    }
  }]);

  return DragPanHandler;
}();

export { DragPanHandler as default };
