// @antv/l7-map/es/handler/shim/drag_rotate.js
import _classCallCheck from "@babel/runtime/helpers/classCallCheck";
import _createClass from "@babel/runtime/helpers/createClass";

var DragRotateHandler = function () {
  function DragRotateHandler(options, mouseRotate, mousePitch) {
    _classCallCheck(this, DragRotateHandler);

    this.mouseRotate = void 0;
    this.mousePitch = void 0;
    this.pitchWithRotate = void 0;
    this.pitchWithRotate = options.pitchWithRotate;
    this.mouseRotate = mouseRotate;
    this.mousePitch = mousePitch;
  }

  _createClass(DragRotateHandler, [{
    key: "enable",
    value: function enable() {
      this.mouseRotate.enable();

      if (this.pitchWithRotate) {
        this.mousePitch.enable();
      }
    }
  }, {
    key: "disable",
    value: function disable() {
      this.mouseRotate.disable();
      this.mousePitch.disable();
    }
  }, {
    key: "isEnabled",
    value: function isEnabled() {
      return this.mouseRotate.isEnabled() && (!this.pitchWithRotate || this.mousePitch.isEnabled());
    }
  }, {
    key: "isActive",
    value: function isActive() {
      return this.mouseRotate.isActive() || this.mousePitch.isActive();
    }
  }]);

  return DragRotateHandler;
}();

export { DragRotateHandler as default };
