// @antv/l7-map/lib/handler/tap/tap_drag_zoom.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _single_tap_recognizer = require("./single_tap_recognizer");

var _tap_recognizer = _interopRequireDefault(require("./tap_recognizer"));

var TapDragZoomHandler = function () {
  function TapDragZoomHandler() {
    (0, _classCallCheck2.default)(this, TapDragZoomHandler);
    this.enabled = void 0;
    this.active = void 0;
    this.swipePoint = void 0;
    this.swipeTouch = void 0;
    this.tapTime = void 0;
    this.tap = void 0;
    this.tap = new _tap_recognizer.default({
      numTouches: 1,
      numTaps: 1
    });
    this.reset();
  }

  (0, _createClass2.default)(TapDragZoomHandler, [{
    key: "reset",
    value: function reset() {
      this.active = false;
      delete this.swipePoint;
      delete this.swipeTouch;
      delete this.tapTime;
      this.tap.reset();
    }
  }, {
    key: "touchstart",
    value: function touchstart(e, points, mapTouches) {
      if (this.swipePoint) {
        return;
      }

      if (this.tapTime && e.timeStamp - this.tapTime > _single_tap_recognizer.MAX_TAP_INTERVAL) {
        this.reset();
      }

      if (!this.tapTime) {
        this.tap.touchstart(e, points, mapTouches);
      } else if (mapTouches.length > 0) {
        this.swipePoint = points[0];
        this.swipeTouch = mapTouches[0].identifier;
      }
    }
  }, {
    key: "touchmove",
    value: function touchmove(e, points, mapTouches) {
      if (!this.tapTime) {
        this.tap.touchmove(e, points, mapTouches);
      } else if (this.swipePoint) {
        if (mapTouches[0].identifier !== this.swipeTouch) {
          return;
        }

        var newSwipePoint = points[0];
        var dist = newSwipePoint.y - this.swipePoint.y;
        this.swipePoint = newSwipePoint;
        e.preventDefault();
        this.active = true;
        return {
          zoomDelta: dist / 128
        };
      }
    }
  }, {
    key: "touchend",
    value: function touchend(e, points, mapTouches) {
      if (!this.tapTime) {
        var point = this.tap.touchend(e, points, mapTouches);

        if (point) {
          this.tapTime = e.timeStamp;
        }
      } else if (this.swipePoint) {
        if (mapTouches.length === 0) {
          this.reset();
        }
      }
    }
  }, {
    key: "touchcancel",
    value: function touchcancel() {
      this.reset();
    }
  }, {
    key: "enable",
    value: function enable() {
      this.enabled = true;
    }
  }, {
    key: "disable",
    value: function disable() {
      this.enabled = false;
      this.reset();
    }
  }, {
    key: "isEnabled",
    value: function isEnabled() {
      return this.enabled;
    }
  }, {
    key: "isActive",
    value: function isActive() {
      return this.active;
    }
  }]);
  return TapDragZoomHandler;
}();

exports.default = TapDragZoomHandler;
