// @antv/l7-map/es/handler/click_zoom.js
import _classCallCheck from "@babel/runtime/helpers/classCallCheck";
import _createClass from "@babel/runtime/helpers/createClass";

var ClickZoomHandler = function () {
  function ClickZoomHandler() {
    _classCallCheck(this, ClickZoomHandler);

    this.enabled = void 0;
    this.active = void 0;
    this.reset();
  }

  _createClass(ClickZoomHandler, [{
    key: "reset",
    value: function reset() {
      this.active = false;
    }
  }, {
    key: "dblclick",
    value: function dblclick(e, point) {
      e.preventDefault();
      return {
        cameraAnimation: function cameraAnimation(map) {
          map.easeTo({
            duration: 300,
            zoom: map.getZoom() + (e.shiftKey ? -1 : 1),
            around: map.unproject(point)
          }, {
            originalEvent: e
          });
        }
      };
    }
  }, {
    key: "enable",
    value: function enable() {
      this.enabled = true;
    }
  }, {
    key: "disable",
    value: function disable() {
      this.enabled = false;
      this.reset();
    }
  }, {
    key: "isEnabled",
    value: function isEnabled() {
      return this.enabled;
    }
  }, {
    key: "isActive",
    value: function isActive() {
      return this.active;
    }
  }]);

  return ClickZoomHandler;
}();

export { ClickZoomHandler as default };
