// @antv/l7-map/es/handler/events/map_touch_event.js
import _classCallCheck from "@babel/runtime/helpers/classCallCheck";
import _createClass from "@babel/runtime/helpers/createClass";
import _inherits from "@babel/runtime/helpers/inherits";
import _possibleConstructorReturn from "@babel/runtime/helpers/possibleConstructorReturn";
import _getPrototypeOf from "@babel/runtime/helpers/getPrototypeOf";

function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = _getPrototypeOf(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = _getPrototypeOf(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return _possibleConstructorReturn(this, result); }; }

function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Boolean.prototype.valueOf.call(Reflect.construct(Boolean, [], function () {})); return true; } catch (e) { return false; } }

import Point from '../../geo/point';
import DOM from '../../utils/dom';
import { Event } from './event';

var MapTouchEvent = function (_Event) {
  _inherits(MapTouchEvent, _Event);

  var _super = _createSuper(MapTouchEvent);

  function MapTouchEvent(type, map, originalEvent) {
    var _this;

    _classCallCheck(this, MapTouchEvent);

    var touches = type === 'touchend' ? originalEvent.changedTouches : originalEvent.touches;
    var points = DOM.touchPos(map.getCanvasContainer(), touches);
    var lngLats = points.map(function (t) {
      return map.unproject(t);
    });
    var point = points.reduce(function (prev, curr, i, arr) {
      return prev.add(curr.div(arr.length));
    }, new Point(0, 0));
    var lngLat = map.unproject(point);
    _this = _super.call(this, type, {
      points: points,
      point: point,
      lngLats: lngLats,
      lngLat: lngLat,
      originalEvent: originalEvent
    });
    _this.type = void 0;
    _this.target = void 0;
    _this.originalEvent = void 0;
    _this.lngLat = void 0;
    _this.point = void 0;
    _this.points = void 0;
    _this.lngLats = void 0;
    _this.defaultPrevented = void 0;
    _this.defaultPrevented = false;
    return _this;
  }

  _createClass(MapTouchEvent, [{
    key: "preventDefault",
    value: function preventDefault() {
      this.defaultPrevented = true;
    }
  }]);

  return MapTouchEvent;
}(Event);

export { MapTouchEvent as default };
