// @antv/l7-core/es/services/component/PopupService.js
import _classCallCheck from "@babel/runtime/helpers/classCallCheck";
import _createClass from "@babel/runtime/helpers/createClass";

var _dec, _class, _temp;

import { injectable } from 'inversify';
import { TYPES } from '../../types';
var PopupService = (_dec = injectable(), _dec(_class = (_temp = function () {
  function PopupService() {
    _classCallCheck(this, PopupService);

    this.scene = void 0;
    this.popup = void 0;
    this.mapsService = void 0;
    this.unAddPopup = void 0;
  }

  _createClass(PopupService, [{
    key: "removePopup",
    value: function removePopup(popup) {
      popup.remove();
    }
  }, {
    key: "destroy",
    value: function destroy() {
      this.popup.remove();
    }
  }, {
    key: "addPopup",
    value: function addPopup(popup) {
      if (this.popup) {
        this.popup.remove();
      }

      if (this.mapsService.map && this.mapsService.getMarkerContainer()) {
        popup.addTo(this.scene);
        this.popup = popup;
      } else {
        this.unAddPopup = popup;
      }
    }
  }, {
    key: "initPopup",
    value: function initPopup() {
      if (this.unAddPopup) {
        this.addPopup(this.unAddPopup);
        this.unAddPopup = null;
      }
    }
  }, {
    key: "init",
    value: function init(scene) {
      this.scene = scene;
      this.mapsService = scene.get(TYPES.IMapService);
    }
  }]);

  return PopupService;
}(), _temp)) || _class);
export { PopupService as default };
