// @antv/l7-core/lib/services/component/PopupService.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _inversify = require("inversify");

var _types = require("../../types");

var _dec, _class, _temp;

var PopupService = (_dec = (0, _inversify.injectable)(), _dec(_class = (_temp = function () {
  function PopupService() {
    (0, _classCallCheck2.default)(this, PopupService);
    this.scene = void 0;
    this.popup = void 0;
    this.mapsService = void 0;
    this.unAddPopup = void 0;
  }

  (0, _createClass2.default)(PopupService, [{
    key: "removePopup",
    value: function removePopup(popup) {
      popup.remove();
    }
  }, {
    key: "destroy",
    value: function destroy() {
      this.popup.remove();
    }
  }, {
    key: "addPopup",
    value: function addPopup(popup) {
      if (this.popup) {
        this.popup.remove();
      }

      if (this.mapsService.map && this.mapsService.getMarkerContainer()) {
        popup.addTo(this.scene);
        this.popup = popup;
      } else {
        this.unAddPopup = popup;
      }
    }
  }, {
    key: "initPopup",
    value: function initPopup() {
      if (this.unAddPopup) {
        this.addPopup(this.unAddPopup);
        this.unAddPopup = null;
      }
    }
  }, {
    key: "init",
    value: function init(scene) {
      this.scene = scene;
      this.mapsService = scene.get(_types.TYPES.IMapService);
    }
  }]);
  return PopupService;
}(), _temp)) || _class);
exports.default = PopupService;
