// @umijs/preset-built-in/lib/plugins/features/fastRefresh.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

function _react() {
  const data = _interopRequireDefault(require("react"));

  _react = function _react() {
    return data;
  };

  return data;
}

function _umi() {
  const data = require("umi");

  _umi = function _umi() {
    return data;
  };

  return data;
}

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

const createDebug = _umi().utils.createDebug;

const debug = createDebug('umi:preset-build-in:fastRefresh');

var _default = api => {
  /**
   * enable by default, back up using view rerender
   * ssr can't work with fastRefresh
   */
  api.describe({
    key: 'fastRefresh',
    config: {
      schema(joi) {
        return joi.object();
      }

    },
    enableBy: api.EnableBy.config
  });
  api.chainWebpack((memo, {
    type
  }) => {
    // must add api.env, test env needed.
    if (api.env === 'development' && type === _umi().BundlerConfigType.csr) {
      memo.plugin('fastRefresh').after('hmr').use(require('@pmmmwh/react-refresh-webpack-plugin'), [{
        overlay: false
      }]);
      debug('FastRefresh webpack loaded');
    }

    return memo;
  });
  api.modifyBabelOpts({
    fn: (babelOpts, {
      type
    }) => {
      if (api.env === 'development' && type === _umi().BundlerConfigType.csr) {
        babelOpts.plugins.push([require.resolve('react-refresh/babel')]);
        debug('FastRefresh babel loaded');
      }

      return babelOpts;
    },
    stage: Infinity
  });
};

exports.default = _default;
