// gg-editor/cjs/components/ContextMenu/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

require("core-js/modules/es.array.concat");

require("core-js/modules/es.object.define-property");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = exports.CanvasMenu = exports.MultiMenu = exports.GroupMenu = exports.EdgeMenu = exports.NodeMenu = void 0;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf3 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _react = _interopRequireDefault(require("react"));

var _utils = require("../../utils");

var _Editor = _interopRequireDefault(require("../Base/Editor"));

var _constants = require("../../common/constants");

var _withGGEditorContext = _interopRequireDefault(require("../../common/context/GGEditorContext/withGGEditorContext"));

var _Menu = _interopRequireDefault(require("./Menu"));

var ContextMenu =
/*#__PURE__*/
function (_React$Component) {
  (0, _inherits2["default"])(ContextMenu, _React$Component);

  function ContextMenu() {
    var _getPrototypeOf2;

    var _this;

    (0, _classCallCheck2["default"])(this, ContextMenu);

    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    _this = (0, _possibleConstructorReturn2["default"])(this, (_getPrototypeOf2 = (0, _getPrototypeOf3["default"])(ContextMenu)).call.apply(_getPrototypeOf2, [this].concat(args)));
    _this.contextMenu = null;
    return _this;
  }

  (0, _createClass2["default"])(ContextMenu, [{
    key: "componentDidMount",
    value: function componentDidMount() {
      var editor = this.props.editor;
      this.contextMenu = new _Editor["default"].Contextmenu({
        container: this.containerId
      });
      editor.add(this.contextMenu);
    }
  }, {
    key: "render",
    value: function render() {
      var children = this.props.children;
      return _react["default"].createElement("div", (0, _extends2["default"])({
        id: this.containerId
      }, (0, _utils.pick)(this.props, ['style', 'className'])), children);
    }
  }, {
    key: "containerId",
    get: function get() {
      var editor = this.props.editor;
      return "".concat(_constants.CONTEXT_MENU_CONTAINER, "_").concat(editor.id);
    }
  }]);
  return ContextMenu;
}(_react["default"].Component);

var NodeMenu = _Menu["default"].create('node');

exports.NodeMenu = NodeMenu;

var EdgeMenu = _Menu["default"].create('edge');

exports.EdgeMenu = EdgeMenu;

var GroupMenu = _Menu["default"].create('group');

exports.GroupMenu = GroupMenu;

var MultiMenu = _Menu["default"].create('multi');

exports.MultiMenu = MultiMenu;

var CanvasMenu = _Menu["default"].create('canvas');

exports.CanvasMenu = CanvasMenu;

var _default = (0, _withGGEditorContext["default"])(ContextMenu);

exports["default"] = _default;
