// @umijs/babel-plugin-import-to-await-require/lib/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.specifiersToProperties = specifiersToProperties;
exports.default = _default;

function _react() {
  const data = _interopRequireDefault(require("react"));

  _react = function _react() {
    return data;
  };

  return data;
}

function t() {
  const data = _interopRequireWildcard(require("@babel/types"));

  t = function t() {
    return data;
  };

  return data;
}

function _getRequireWildcardCache() { if (typeof WeakMap !== "function") return null; var cache = new WeakMap(); _getRequireWildcardCache = function _getRequireWildcardCache() { return cache; }; return cache; }

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function specifiersToProperties(specifiers) {
  return specifiers.reduce((memo, s) => {
    if (t().isImportDefaultSpecifier(s)) {
      memo.properties.push(t().objectProperty(t().identifier('default'), s.local));
    } else if (t().isExportDefaultSpecifier(s)) {
      memo.properties.push(t().objectProperty(t().identifier('default'), s.exported));
    } else if (t().isExportSpecifier(s)) {
      memo.properties.push(t().objectProperty(s.local, s.exported));
    } else if (t().isImportNamespaceSpecifier(s)) {
      memo.namespaceIdentifier = s.local;
    } else {
      memo.properties.push(t().objectProperty(s.imported, s.local));
    }

    return memo;
  }, {
    properties: [],
    namespaceIdentifier: null
  });
}

function isMatchLib(path, libs, alias) {
  return libs.concat(Object.keys(alias)).some(lib => {
    if (typeof lib === 'string') {
      return lib === path;
    } else {
      return lib.test(path);
    }
  });
}

function getPath(path, alias) {
  const keys = Object.keys(alias);

  for (var _i = 0, _keys = keys; _i < _keys.length; _i++) {
    const key = _keys[_i];

    if (path.startsWith(key)) {
      return path.replace(key, alias[key]);
    }
  }

  return path;
}

function _default() {
  return {
    visitor: {
      Program: {
        exit(path, {
          opts
        }) {
          const variableDeclarations = [];
          let index = path.node.body.length - 1;

          while (index >= 0) {
            const d = path.node.body[index];

            if (t().isImportDeclaration(d)) {
              var _opts$onTransformDeps;

              const isMatch = isMatchLib(d.source.value, opts.libs, opts.alias || {});
              (_opts$onTransformDeps = opts.onTransformDeps) === null || _opts$onTransformDeps === void 0 ? void 0 : _opts$onTransformDeps.call(opts, {
                source: d.source.value,
                file: path.hub.file.opts.filename,
                isMatch
              });

              if (isMatch) {
                const _specifiersToProperti = specifiersToProperties(d.specifiers),
                      properties = _specifiersToProperti.properties,
                      namespaceIdentifier = _specifiersToProperti.namespaceIdentifier;

                const id = t().objectPattern(properties);
                const init = t().awaitExpression(t().callExpression(t().import(), [t().stringLiteral(`${opts.remoteName}/${getPath(d.source.value, opts.alias || {})}`)]));

                if (namespaceIdentifier) {
                  if (properties.length) {
                    variableDeclarations.unshift(t().variableDeclaration('const', [t().variableDeclarator(id, namespaceIdentifier)]));
                  }

                  variableDeclarations.unshift(t().variableDeclaration('const', [t().variableDeclarator(namespaceIdentifier, init)]));
                } else {
                  variableDeclarations.unshift(t().variableDeclaration('const', [t().variableDeclarator(id, init)]));
                }

                path.node.body.splice(index, 1);
              }
            }

            if (t().isExportAllDeclaration(d) && d.source) {
              var _opts$onTransformDeps2;

              (_opts$onTransformDeps2 = opts.onTransformDeps) === null || _opts$onTransformDeps2 === void 0 ? void 0 : _opts$onTransformDeps2.call(opts, {
                source: d.source.value,
                file: path.hub.file.opts.filename,
                isMatch: false,
                isExportAllDeclaration: true
              });
            } // export { bar } from 'foo';


            if (t().isExportNamedDeclaration(d) && d.source) {
              var _opts$onTransformDeps3;

              const isMatch = isMatchLib(d.source.value, opts.libs, opts.alias || {});
              (_opts$onTransformDeps3 = opts.onTransformDeps) === null || _opts$onTransformDeps3 === void 0 ? void 0 : _opts$onTransformDeps3.call(opts, {
                source: d.source.value,
                file: path.hub.file.opts.filename,
                isMatch
              });

              if (isMatch) {
                const _specifiersToProperti2 = specifiersToProperties(d.specifiers),
                      properties = _specifiersToProperti2.properties;

                const id = t().objectPattern(properties);
                const init = t().awaitExpression(t().callExpression(t().import(), [t().stringLiteral(`${opts.remoteName}/${getPath(d.source.value, opts.alias || {})}`)]));
                variableDeclarations.unshift(t().variableDeclaration('const', [t().variableDeclarator(id, init)]));
                d.source = null;
              }
            }

            index -= 1;
          }

          path.node.body = [...variableDeclarations, ...path.node.body];
        }

      }
    }
  };
}
