// gg-editor/es/components/Koni/index.js
import "core-js/modules/es.array.concat";
import _classCallCheck from "@babel/runtime/helpers/classCallCheck";
import _createClass from "@babel/runtime/helpers/createClass";
import _possibleConstructorReturn from "@babel/runtime/helpers/possibleConstructorReturn";
import _getPrototypeOf from "@babel/runtime/helpers/getPrototypeOf";
import _inherits from "@babel/runtime/helpers/inherits";
import Editor from "../Base/Editor";
import { KONI_CONTAINER, KONI_CLASS_NAME, EVENT_BEFORE_ADD_PAGE, EVENT_AFTER_ADD_PAGE } from "../../common/constants";
import Page from "../Page";
import withGGEditorContext from "../../common/context/GGEditorContext/withGGEditorContext";

var Koni =
/*#__PURE__*/
function (_Page) {
  _inherits(Koni, _Page);

  function Koni() {
    _classCallCheck(this, Koni);

    return _possibleConstructorReturn(this, _getPrototypeOf(Koni).apply(this, arguments));
  }

  _createClass(Koni, [{
    key: "initPage",
    value: function initPage() {
      var editor = this.props.editor;
      editor.emit(EVENT_BEFORE_ADD_PAGE, {
        className: KONI_CLASS_NAME
      });
      this.page = new Editor.Koni(this.config);
      editor.add(this.page);
      editor.emit(EVENT_AFTER_ADD_PAGE, {
        page: this.page
      });
    }
  }, {
    key: "pageId",
    get: function get() {
      var editor = this.props.editor;
      return "".concat(KONI_CONTAINER, "_").concat(editor.id);
    }
  }]);

  return Koni;
}(Page);

Koni.defaultProps = {
  data: {
    nodes: [],
    edges: []
  }
};
export default withGGEditorContext(Koni);
