// @antv/l7-map/lib/geo/simple.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.mercatorXfromLng = mercatorXfromLng;
exports.mercatorYfromLat = mercatorYfromLat;
exports.mercatorZfromAltitude = mercatorZfromAltitude;
exports.lngFromMercatorX = lngFromMercatorX;
exports.latFromMercatorY = latFromMercatorY;
exports.altitudeFromMercatorZ = altitudeFromMercatorZ;
exports.mercatorScale = mercatorScale;
exports.default = void 0;

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _lng_lat = _interopRequireWildcard(require("../geo/lng_lat"));

var earthCircumfrence = 2 * Math.PI * _lng_lat.earthRadius;

function circumferenceAtLatitude(latitude) {
  return earthCircumfrence * Math.cos(latitude * Math.PI / 180);
}

function mercatorXfromLng(lng) {
  return lng;
}

function mercatorYfromLat(lat) {
  return lat;
}

function mercatorZfromAltitude(altitude, lat) {
  return altitude;
}

function lngFromMercatorX(x) {
  return x;
}

function latFromMercatorY(y) {
  return y;
}

function altitudeFromMercatorZ(z, y) {
  return z;
}

function mercatorScale(lat) {
  return 1;
}

var SimpleCoordinate = function () {
  function SimpleCoordinate(x, y) {
    var z = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : 0;
    (0, _classCallCheck2.default)(this, SimpleCoordinate);
    this.x = void 0;
    this.y = void 0;
    this.z = void 0;
    this.x = +x;
    this.y = +y;
    this.z = +z;
  }

  (0, _createClass2.default)(SimpleCoordinate, [{
    key: "toLngLat",
    value: function toLngLat() {
      return new _lng_lat.default(this.x, this.y);
    }
  }, {
    key: "toAltitude",
    value: function toAltitude() {
      return this.z;
    }
  }, {
    key: "meterInMercatorCoordinateUnits",
    value: function meterInMercatorCoordinateUnits() {
      return 1;
    }
  }], [{
    key: "fromLngLat",
    value: function fromLngLat(lngLatLike) {
      var altitude = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : 0;

      var lngLat = _lng_lat.default.convert(lngLatLike);

      return new SimpleCoordinate(mercatorXfromLng(lngLat.lng), mercatorYfromLat(lngLat.lat), mercatorZfromAltitude(altitude, lngLat.lat));
    }
  }]);
  return SimpleCoordinate;
}();

exports.default = SimpleCoordinate;
