// gg-editor/cjs/components/GGEditor/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

require("core-js/modules/es.array.for-each");

require("core-js/modules/es.object.define-property");

require("core-js/modules/web.dom-collections.for-each");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = void 0;

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _assertThisInitialized2 = _interopRequireDefault(require("@babel/runtime/helpers/assertThisInitialized"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _react = _interopRequireDefault(require("react"));

var _Editor = _interopRequireDefault(require("../Base/Editor"));

var _constants = require("../../common/constants");

var _utils = require("../../utils");

var _Global = _interopRequireDefault(require("../../common/Global"));

var _GGEditorContext = _interopRequireDefault(require("../../common/context/GGEditorContext"));

var _PropsAPIContext = _interopRequireDefault(require("../../common/context/PropsAPIContext"));

var _propsAPI = _interopRequireDefault(require("../../common/context/PropsAPIContext/propsAPI"));

var GGEditor =
/*#__PURE__*/
function (_React$Component) {
  (0, _inherits2["default"])(GGEditor, _React$Component);
  (0, _createClass2["default"])(GGEditor, [{
    key: "currentPage",
    get: function get() {
      return this.editor.getCurrentPage();
    }
  }], [{
    key: "setTrackable",
    value: function setTrackable(value) {
      _Global["default"].set('trackable', Boolean(value));
    }
  }]);

  function GGEditor(props) {
    var _this;

    (0, _classCallCheck2["default"])(this, GGEditor);
    _this = (0, _possibleConstructorReturn2["default"])(this, (0, _getPrototypeOf2["default"])(GGEditor).call(this, props));
    _this.editor = null;

    _this.addListener = function (target, eventName, handler) {
      if (typeof handler === 'function') target.on(eventName, handler);
    };

    _this.handleBeforeAddPage = function (func) {
      _this.editor.on(_constants.EVENT_BEFORE_ADD_PAGE, func);
    };

    _this.handleAfterAddPage = function (func) {
      var _assertThisInitialize = (0, _assertThisInitialized2["default"])(_this),
          page = _assertThisInitialize.currentPage;

      if (page) {
        func({
          page: page
        });
        return;
      }

      _this.editor.on(_constants.EVENT_AFTER_ADD_PAGE, func);
    };

    _this.init();

    _this.bindEvent();

    return _this;
  }

  (0, _createClass2["default"])(GGEditor, [{
    key: "init",
    value: function init() {
      this.editor = new _Editor["default"]();
      this.ggEditor = {
        editor: this.editor,
        onBeforeAddPage: this.handleBeforeAddPage,
        onAfterAddPage: this.handleAfterAddPage
      };
      this.propsAPI = new _propsAPI["default"](this.editor);
    }
  }, {
    key: "bindEvent",
    value: function bindEvent() {
      var _this2 = this;

      _constants.EDITOR_EVENTS.forEach(function (event) {
        _this2.addListener(_this2.editor, [event], _this2.props[_constants.EDITOR_REACT_EVENTS[event]]);
      });
    }
  }, {
    key: "componentWillUnmount",
    value: function componentWillUnmount() {
      this.editor.destroy();
    }
  }, {
    key: "render",
    value: function render() {
      var children = this.props.children;
      return _react["default"].createElement(_GGEditorContext["default"].Provider, {
        value: this.ggEditor
      }, _react["default"].createElement(_PropsAPIContext["default"].Provider, {
        value: this.propsAPI
      }, _react["default"].createElement("div", (0, _utils.pick)(this.props, ['style', 'className']), children)));
    }
  }]);
  return GGEditor;
}(_react["default"].Component);

var _default = GGEditor;
exports["default"] = _default;
