// @antv/l7-map/lib/handler/shim/dblclick_zoom.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var DoubleClickZoomHandler = function () {
  function DoubleClickZoomHandler(clickZoom, TapZoom) {
    (0, _classCallCheck2.default)(this, DoubleClickZoomHandler);
    this.clickZoom = void 0;
    this.tapZoom = void 0;
    this.clickZoom = clickZoom;
    this.tapZoom = TapZoom;
  }

  (0, _createClass2.default)(DoubleClickZoomHandler, [{
    key: "enable",
    value: function enable() {
      this.clickZoom.enable();
      this.tapZoom.enable();
    }
  }, {
    key: "disable",
    value: function disable() {
      this.clickZoom.disable();
      this.tapZoom.disable();
    }
  }, {
    key: "isEnabled",
    value: function isEnabled() {
      return this.clickZoom.isEnabled() && this.tapZoom.isEnabled();
    }
  }, {
    key: "isActive",
    value: function isActive() {
      return this.clickZoom.isActive() || this.tapZoom.isActive();
    }
  }]);
  return DoubleClickZoomHandler;
}();

exports.default = DoubleClickZoomHandler;
