// @antv/l7-map/lib/handler/shim/drag_pan.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var DragPanHandler = function () {
  function DragPanHandler(el, mousePan, touchPan) {
    (0, _classCallCheck2.default)(this, DragPanHandler);
    this.inertiaOptions = void 0;
    this.el = void 0;
    this.mousePan = void 0;
    this.touchPan = void 0;
    this.el = el;
    this.mousePan = mousePan;
    this.touchPan = touchPan;
  }

  (0, _createClass2.default)(DragPanHandler, [{
    key: "enable",
    value: function enable(options) {
      this.inertiaOptions = options || {};
      this.mousePan.enable();
      this.touchPan.enable();
      this.el.classList.add('l7-touch-drag-pan');
    }
  }, {
    key: "disable",
    value: function disable() {
      this.mousePan.disable();
      this.touchPan.disable();
      this.el.classList.remove('l7-touch-drag-pan');
    }
  }, {
    key: "isEnabled",
    value: function isEnabled() {
      return this.mousePan.isEnabled() && this.touchPan.isEnabled();
    }
  }, {
    key: "isActive",
    value: function isActive() {
      return this.mousePan.isActive() || this.touchPan.isActive();
    }
  }]);
  return DragPanHandler;
}();

exports.default = DragPanHandler;
