// @antv/l7-core/lib/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

Object.defineProperty(exports, "__esModule", {
  value: true
});
var _exportNames = {
  container: true,
  createLayerContainer: true,
  createSceneContainer: true,
  lazyInject: true,
  lazyMultiInject: true,
  BasePostProcessingPass: true,
  TYPES: true,
  packCircleVertex: true
};
Object.defineProperty(exports, "container", {
  enumerable: true,
  get: function get() {
    return _inversify.default;
  }
});
Object.defineProperty(exports, "createLayerContainer", {
  enumerable: true,
  get: function get() {
    return _inversify.createLayerContainer;
  }
});
Object.defineProperty(exports, "createSceneContainer", {
  enumerable: true,
  get: function get() {
    return _inversify.createSceneContainer;
  }
});
Object.defineProperty(exports, "lazyInject", {
  enumerable: true,
  get: function get() {
    return _inversify.lazyInject;
  }
});
Object.defineProperty(exports, "lazyMultiInject", {
  enumerable: true,
  get: function get() {
    return _inversify.lazyMultiInject;
  }
});
Object.defineProperty(exports, "BasePostProcessingPass", {
  enumerable: true,
  get: function get() {
    return _BasePostProcessingPass.default;
  }
});
Object.defineProperty(exports, "TYPES", {
  enumerable: true,
  get: function get() {
    return _types.TYPES;
  }
});
Object.defineProperty(exports, "packCircleVertex", {
  enumerable: true,
  get: function get() {
    return _vertexCompression.packCircleVertex;
  }
});

var _inversify = _interopRequireWildcard(require("./inversify.config"));

var _BasePostProcessingPass = _interopRequireDefault(require("./services/renderer/passes/BasePostProcessingPass"));

var _types = require("./types");

var _vertexCompression = require("./utils/vertex-compression");

var _ILayerService = require("./services/layer/ILayerService");

Object.keys(_ILayerService).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _ILayerService[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function get() {
      return _ILayerService[key];
    }
  });
});

var _IStyleAttributeService = require("./services/layer/IStyleAttributeService");

Object.keys(_IStyleAttributeService).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _IStyleAttributeService[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function get() {
      return _IStyleAttributeService[key];
    }
  });
});

var _ISourceService = require("./services/source/ISourceService");

Object.keys(_ISourceService).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _ISourceService[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function get() {
      return _ISourceService[key];
    }
  });
});

var _IMapService = require("./services/map/IMapService");

Object.keys(_IMapService).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _IMapService[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function get() {
      return _IMapService[key];
    }
  });
});

var _ICoordinateSystemService = require("./services/coordinate/ICoordinateSystemService");

Object.keys(_ICoordinateSystemService).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _ICoordinateSystemService[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function get() {
      return _ICoordinateSystemService[key];
    }
  });
});

var _IRendererService = require("./services/renderer/IRendererService");

Object.keys(_IRendererService).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _IRendererService[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function get() {
      return _IRendererService[key];
    }
  });
});

var _ICameraService = require("./services/camera/ICameraService");

Object.keys(_ICameraService).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _ICameraService[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function get() {
      return _ICameraService[key];
    }
  });
});

var _IConfigService = require("./services/config/IConfigService");

Object.keys(_IConfigService).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _IConfigService[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function get() {
      return _IConfigService[key];
    }
  });
});

var _ISceneService = require("./services/scene/ISceneService");

Object.keys(_ISceneService).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _ISceneService[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function get() {
      return _ISceneService[key];
    }
  });
});

var _IShaderModuleService = require("./services/shader/IShaderModuleService");

Object.keys(_IShaderModuleService).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _IShaderModuleService[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function get() {
      return _IShaderModuleService[key];
    }
  });
});

var _IIconService = require("./services/asset/IIconService");

Object.keys(_IIconService).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _IIconService[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function get() {
      return _IIconService[key];
    }
  });
});

var _IFontService = require("./services/asset/IFontService");

Object.keys(_IFontService).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _IFontService[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function get() {
      return _IFontService[key];
    }
  });
});

var _IControlService = require("./services/component/IControlService");

Object.keys(_IControlService).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _IControlService[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function get() {
      return _IControlService[key];
    }
  });
});

var _IMarkerService = require("./services/component/IMarkerService");

Object.keys(_IMarkerService).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _IMarkerService[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function get() {
      return _IMarkerService[key];
    }
  });
});

var _IPopupService = require("./services/component/IPopupService");

Object.keys(_IPopupService).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _IPopupService[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function get() {
      return _IPopupService[key];
    }
  });
});

var _ILogService = require("./services/log/ILogService");

Object.keys(_ILogService).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _ILogService[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function get() {
      return _ILogService[key];
    }
  });
});

var _IInteractionService = require("./services/interaction/IInteractionService");

Object.keys(_IInteractionService).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _IInteractionService[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function get() {
      return _IInteractionService[key];
    }
  });
});

var _IPickingService = require("./services/interaction/IPickingService");

Object.keys(_IPickingService).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _IPickingService[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function get() {
      return _IPickingService[key];
    }
  });
});

var _IAttribute = require("./services/renderer/IAttribute");

Object.keys(_IAttribute).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _IAttribute[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function get() {
      return _IAttribute[key];
    }
  });
});

var _IBuffer = require("./services/renderer/IBuffer");

Object.keys(_IBuffer).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _IBuffer[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function get() {
      return _IBuffer[key];
    }
  });
});

var _IElements = require("./services/renderer/IElements");

Object.keys(_IElements).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _IElements[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function get() {
      return _IElements[key];
    }
  });
});

var _IFramebuffer = require("./services/renderer/IFramebuffer");

Object.keys(_IFramebuffer).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _IFramebuffer[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function get() {
      return _IFramebuffer[key];
    }
  });
});

var _IModel = require("./services/renderer/IModel");

Object.keys(_IModel).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _IModel[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function get() {
      return _IModel[key];
    }
  });
});

var _IMultiPassRenderer = require("./services/renderer/IMultiPassRenderer");

Object.keys(_IMultiPassRenderer).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _IMultiPassRenderer[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function get() {
      return _IMultiPassRenderer[key];
    }
  });
});

var _IRenderbuffer = require("./services/renderer/IRenderbuffer");

Object.keys(_IRenderbuffer).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _IRenderbuffer[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function get() {
      return _IRenderbuffer[key];
    }
  });
});

var _ITexture2D = require("./services/renderer/ITexture2D");

Object.keys(_ITexture2D).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _ITexture2D[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function get() {
      return _ITexture2D[key];
    }
  });
});

var _IUniform = require("./services/renderer/IUniform");

Object.keys(_IUniform).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _IUniform[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function get() {
      return _IUniform[key];
    }
  });
});

var _gl = require("./services/renderer/gl");

Object.keys(_gl).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _gl[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function get() {
      return _gl[key];
    }
  });
});
